server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java

```diff
@@ -0,0 +1,215 @@
+import static java.util.stream.Collectors.toMap;
+import static org.sonar.api.measures.CoreMetrics.NEW_LINES_DUPLICATED_KEY;
+import static org.sonar.server.computation.task.projectanalysis.period.PeriodPredicates.viewsRestrictedPeriods;
+
+/**
+ * Computes new duplication measures on files and then aggregates them on higher components.
+ *
+ */
+public class NewDuplicationMeasuresStep implements ComputationStep {
+
+  private final ImmutableList<Formula> formulas;
+
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodsHolder periodsHolder;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  public NewDuplicationMeasuresStep(TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder, MetricRepository metricRepository, MeasureRepository measureRepository,
+    ScmInfoRepository scmInfoRepository, @Nullable DuplicationRepository duplicationRepository) {
+    this.treeRootHolder = treeRootHolder;
+    this.periodsHolder = periodsHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+    this.formulas = ImmutableList.of(NewDuplicationFormula.from(scmInfoRepository, duplicationRepository));
+  }
+
+  /**
+   * Constructor used by Pico in Governance where no DuplicationRepository is available.
+   */
+  public NewDuplicationMeasuresStep(TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder, MetricRepository metricRepository, MeasureRepository measureRepository) {
+    this(treeRootHolder, periodsHolder, metricRepository, measureRepository, null, null);
+  }
+
+  @Override
+  public String getDescription() {
+    return "Compute new duplication measures";
+  }
+
+  @Override
+  public void execute() {
+    new PathAwareCrawler<>(
+      FormulaExecutorComponentVisitor.newBuilder(metricRepository, measureRepository)
+        .withVariationSupport(periodsHolder)
+        .buildFor(formulas))
+          .visit(treeRootHolder.getRoot());
+  }
+
+  private static class NewDuplicationCounter implements Counter<NewDuplicationCounter> {
+    @CheckForNull
+    private final DuplicationRepository duplicationRepository;
+    private final ScmInfoRepository scmInfoRepository;
+    private final IntVariationValue.Array newLines = IntVariationValue.newArray();
+
+    private NewDuplicationCounter(@Nullable DuplicationRepository duplicationRepository, ScmInfoRepository scmInfoRepository) {
+      this.duplicationRepository = duplicationRepository;
+      this.scmInfoRepository = scmInfoRepository;
+    }
+
+    @Override
+    public void aggregate(NewDuplicationCounter counter) {
+      this.newLines.incrementAll(counter.newLines);
+    }
+
+    @Override
+    public void initialize(CounterInitializationContext context) {
+      Component leaf = context.getLeaf();
+      Iterable<Duplication> duplications = requireNonNull(this.duplicationRepository, "DuplicationRepository missing").getDuplications(leaf);
```

Comment:
I don't see any use case when duplicationRepository could be null here, you should remove the requireNonNull